src/sparse_blas/backends/cusparse/operations/cusparse_spmv.cpp

```diff
@@ -0,0 +1,302 @@
+        auto cu_x = x_handle->backend_handle;
+        auto cu_y = y_handle->backend_handle;
+        auto type = A_handle->value_container.data_type;
+        auto cu_op = get_cuda_operation(type, opA);
+        auto cu_type = get_cuda_value_type(type);
+        auto cu_alg = get_cuda_spmv_alg(alg);
+        set_pointer_mode(cu_handle, is_alpha_host_accessible);
+        auto status = cusparseSpMV_bufferSize(cu_handle, cu_op, alpha, cu_a, cu_x, beta, cu_y,
+                                              cu_type, cu_alg, &temp_buffer_size);
+        check_status(status, __func__);
+    };
+    auto event = dispatch_submit(__func__, queue, functor, A_handle, x_handle, y_handle);
+    event.wait_and_throw();
+    spmv_descr->temp_buffer_size = temp_buffer_size;
+    spmv_descr->buffer_size_called = true;
+}
+
+inline void common_spmv_optimize(oneapi::mkl::transpose opA, bool is_alpha_host_accessible,
+                                 matrix_view A_view, matrix_handle_t A_handle,
+                                 dense_vector_handle_t x_handle, bool is_beta_host_accessible,
+                                 dense_vector_handle_t y_handle, spmv_alg alg,
+                                 spmv_descr_t spmv_descr) {
+    check_valid_spmv("spmv_optimize", opA, A_view, A_handle, x_handle, y_handle,
+                     is_alpha_host_accessible, is_beta_host_accessible);
+    if (!spmv_descr->buffer_size_called) {
+        throw mkl::uninitialized("sparse_blas", "spmv_optimize",
+                                 "spmv_buffer_size must be called before spmv_optimize.");
+    }
+    spmv_descr->optimized_called = true;
+    spmv_descr->last_optimized_opA = opA;
+    spmv_descr->last_optimized_A_view = A_view;
+    spmv_descr->last_optimized_A_handle = A_handle;
+    spmv_descr->last_optimized_x_handle = x_handle;
+    spmv_descr->last_optimized_y_handle = y_handle;
+    spmv_descr->last_optimized_alg = alg;
+}
+
+#if CUSPARSE_VERSION >= 12300
+// cusparseSpMV_preprocess was added in cuSPARSE 12.3.0.142 (CUDA 12.4)
+void spmv_optimize_impl(cusparseHandle_t cu_handle, oneapi::mkl::transpose opA, const void *alpha,
+                        matrix_handle_t A_handle, dense_vector_handle_t x_handle, const void *beta,
+                        dense_vector_handle_t y_handle, spmv_alg alg, void *workspace_ptr,
+                        bool is_alpha_host_accessible) {
+    auto cu_a = A_handle->backend_handle;
+    auto cu_x = x_handle->backend_handle;
+    auto cu_y = y_handle->backend_handle;
+    auto type = A_handle->value_container.data_type;
+    auto cu_op = get_cuda_operation(type, opA);
+    auto cu_type = get_cuda_value_type(type);
+    auto cu_alg = get_cuda_spmv_alg(alg);
+    set_pointer_mode(cu_handle, is_alpha_host_accessible);
+    auto status = cusparseSpMV_preprocess(cu_handle, cu_op, alpha, cu_a, cu_x, beta, cu_y, cu_type,
+                                          cu_alg, workspace_ptr);
+    check_status(status, "optimize_spmv");
```

Comment:
```suggestion
    check_status(status, "spmv_optimize");
```

Reply:
Fixed in https://github.com/oneapi-src/oneMKL/pull/527/commits/325f79495c2090b95538e0eaaa80ccf40f2e9737